### PR TITLE
#3358 Adds parallel processing to outlier filter.

### PR DESCRIPTION
**Pull request by whereskenneth**

### CODE REVIEW

File: test/unit/filters/OutlierFilterTest.cpp [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+        outlier_filter.setInput(reader);
+
+        serial_radius_pipeline.prepare();
+        time_point<system_clock> tick = system_clock::now();
+        PointViewSet viewSet =
+            outlier_filter.execute(serial_radius_pipeline.pointTable());
+        time_point<system_clock> tock = system_clock::now();
+
+        auto diff = tock - tick;
+        std::cout
```
Comment (whereskenneth) [inline test/unit/filters/OutlierFilterTest.cpp:241]: Perhaps there is a better option than just dumping to stdout?

File: test/unit/filters/OutlierFilterTest.cpp [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+    ASSERT_EQ(parallel_statistical_outliers->size(), serial_statistical_outliers->size());
+
+}
+
+TEST(OutlierFilterTest, thread_count_validation) {
+    using namespace Dimension;
+    using namespace std::chrono;
+
+    const size_t point_count = 100;
+    auto log = Log::makeLog("outlier_filter_test", "stdout");
```
Comment (whereskenneth) [inline test/unit/filters/OutlierFilterTest.cpp:339]: Is there a better way to check logging?

**Comment by abellgithub** [comment]
This can be (and should) be done lock-free by grouping the points. You can simply pass a start/length to the threads.

**Comment by whereskenneth** [comment]
> This can be (and should) be done lock-free by grouping the points. You can simply pass a start/length to the threads.

Fair enough, but the extra bookkeeping required to implement that isn't worth the readability cost, in my opinion, so I won't author it. Someone else can if they'd like to do it, though.

**Comment by abellgithub** [comment]
? I think you're misunderstanding. I'll do one and if you're interested in helping with others that would be great.

**Comment by abellgithub** [comment]
Maybe I'm missing something, but wouldn't this work without the locks?

```
int numThreads = 4;
ThreadPool pool(numThreads);

    double chunkSize = double(np) / numThreads;
    for (int t = 0; t < numThreads; ++t)
    {
        PointId start = std::floor(chunkSize * t);
        PointId end = std::floor(chunkSize * (t + 1));
        pool.add([&distances, count, &index, start, end]()
        {
            PointIdList indices(count);
            std::vector<double> sqr_dists(count);

            for (PointId i = start; i < end; ++i)
            {
                index.knnSearch(i, count, &indices, &sqr_dists);
                for (size_t j = 1; j < count; ++j)
                {
                    double delta = std::sqrt(sqr_dists[j]) - distances[i];
                    distances[i] += (delta / j);
                }
            }
        });
        start += chunkSize;
    }
    pool.join();
```

Would it be OK to at least take the tests that you've made?

**Comment by whereskenneth** [comment]
Yeah, you can take anything you'd like out of there. 

Are you sure updating `distances[i]` is thread safe? Trying to research but can't find a clear answer.

**Comment by whereskenneth** [comment]
23.2.2  Container data races[container.requirements.dataraces]1For purposes of avoiding data races (17.6.5.9),implementations shall consider the following functions to beconst:begin,end,rbegin,rend,front,back,data,find,lower_bound,upper_bound,equal_range,atand, except in associative or unordered associative containers,operator[].

I guess that means this is fine?

**Comment by abellgithub** [comment]
Yes, direct access to a vector is thread-safe as long as you've not doing any modification to the vector itself (rather than the data, which is fine).

**Comment by hobu** [comment]
Closing this one for now. Please reopen if you pick the effort back up again.

**Pull request closed without merging**